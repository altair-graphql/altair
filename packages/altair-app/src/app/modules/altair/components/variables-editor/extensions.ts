import { jsonSchema } from 'codemirror-json-schema';

export const gqlVariables = () => {
  return [
    // start with an empty schema
    jsonSchema({
      type: 'object',
      properties: {},
    }),
  ];
};
