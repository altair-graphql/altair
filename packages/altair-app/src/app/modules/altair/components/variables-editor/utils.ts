import {
  getNullableType,
  GraphQLInputObjectType,
  GraphQLInputType,
  isEnumType,
  isLeafType,
  isListType,
  isNullableType,
  isSpecifiedScalarType,
} from 'graphql';
import {
  JSONSchema7,
  JSONSchema7Definition,
  JSONSchema7Type,
  JSONSchema7TypeName,
} from 'json-schema';

/**
 * Maximum depth to traverse the GraphQL type tree for variable type resolution.
 * This is to prevent slow performance because of resolving the variable types and/or infinite recursion.
 */
const MAX_DEPTH = 4;

export type VariableToType = Record<string, GraphQLInputType>;

const specifiedScalarTypeToJSONSchemaType: Record<string, JSONSchema7TypeName> =
  {
    String: 'string',
    Int: 'integer',
    Float: 'number',
    Boolean: 'boolean',
    ID: 'string',
  };
const wrapWithAnyOfNull = (def: JSONSchema7, wrap?: boolean) => {
  if (!wrap) {
    return def;
  }
  return {
    anyOf: [def, { type: 'null' as JSONSchema7TypeName }],
  };
};
export const graphqlInputTypeToJsonSchemaType = (
  originalType: GraphQLInputType,
  depth = 0,
  defaultValue?: JSONSchema7Type
): JSONSchema7 => {
  if (depth > MAX_DEPTH) {
    // Max depth reached. Default to string
    return { type: 'string' };
  }

  const type = getNullableType(originalType);

  // Handle enums
  if (isEnumType(type)) {
    return wrapWithAnyOfNull(
      {
        type: 'string',
        description: type.description ?? undefined,
        default: defaultValue,
        anyOf: type.getValues().map((v) => ({
          enum: [v.value],
          description: v.description ?? undefined,
        })),
      },
      isNullableType(originalType)
    );
  }
  if (isLeafType(type)) {
    if (isSpecifiedScalarType(type)) {
      return wrapWithAnyOfNull(
        {
          type:
            specifiedScalarTypeToJSONSchemaType[type.toString()] ?? 'string',
          description: type.description ?? undefined,
          default: defaultValue,
        },
        isNullableType(originalType)
      );
    }
    // custom scalar types default to string
    return wrapWithAnyOfNull(
      {
        type: 'string',
        description: type.description ?? undefined,
      },
      isNullableType(originalType)
    );
  }
  // Handle lists
  if (isListType(type)) {
    return wrapWithAnyOfNull(
      {
        type: 'array',
        items: graphqlInputTypeToJsonSchemaType(type.ofType, depth + 1),
        description: type.ofType?.description ?? undefined,
        default: defaultValue,
      },
      isNullableType(originalType)
    );
  }

  if (type instanceof GraphQLInputObjectType) {
    return {
      type: 'object',
      description: type.description ?? undefined,
      properties: Object.entries(type.getFields()).reduce(
        (acc, [key, field]) => {
          acc[key] = graphqlInputTypeToJsonSchemaType(
            field.type,
            depth + 1,
            field.defaultValue
          );
          return acc;
        },
        {} as Record<string, JSONSchema7Definition>
      ),
    };
  }

  // Unknown case. Default to string
  return { type: 'string' };
};

/*
 * Converts a VariableToType object to a JSON Schema object.
 * This is used to generate a JSON Schema for the variables editor.
 */
export const vttToJsonSchema = (vtt?: VariableToType): JSONSchema7 => {
  if (!vtt) {
    return {
      type: 'object',
      properties: {},
    };
  }
  return {
    type: 'object',
    properties: Object.entries(vtt).reduce((acc, [key, type]) => {
      acc[key] = graphqlInputTypeToJsonSchemaType(type);
      return acc;
    }, {} as Record<string, JSONSchema7Definition>),
  };
};
