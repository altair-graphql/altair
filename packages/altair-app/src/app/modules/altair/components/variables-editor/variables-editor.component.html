@if (!betaDisableNewEditor) {
  <app-codemirror
    #newEditor
    class="set-variable-textarea mousetrap"
    [extensions]="editorExtensions"
    [ngModel]="variables"
    (ngModelChange)="variablesChange.emit($event)"
  ></app-codemirror>
}
@if (betaDisableNewEditor) {
  <ngx-codemirror
    #editor
    class="set-variable-textarea mousetrap"
    [options]="variableEditorConfig"
    [ngModel]="variables"
    (ngModelChange)="variablesChange.emit($event)"
  ></ngx-codemirror>
}
