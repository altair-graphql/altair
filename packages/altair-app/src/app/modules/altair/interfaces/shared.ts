export interface IDictionary<V = any> {
  [key: string]: V;
}

export type Omit<T, K extends keyof T> = Pick<T, Exclude<keyof T, K>>;

export type TODO = any;

export type UnknownError = TODO;

export interface TrackByIdItem {
  id: string;
}
