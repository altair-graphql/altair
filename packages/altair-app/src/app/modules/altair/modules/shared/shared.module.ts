import { NgModule, ModuleWithProviders } from '@angular/core';
import { CommonModule } from '@angular/common';

import { DragDropModule } from '@angular/cdk/drag-drop';
import { TranslateModule } from '@ngx-translate/core';
import { NgPipesModule } from 'ngx-pipes';
import { NguiAutoCompleteModule } from '@ngui/auto-complete';
import { MarkdownModule } from 'ngx-markdown';
import { HotToastModule } from '@ngneat/hot-toast';
import { ResizableModule } from 'angular-resizable-element';

import { NzButtonModule } from 'ng-zorro-antd/button';
import { NzFormModule } from 'ng-zorro-antd/form';
import { NzSelectModule } from 'ng-zorro-antd/select';
import { NzModalModule } from 'ng-zorro-antd/modal';
import { NzDropDownModule } from 'ng-zorro-antd/dropdown';
import { NzSwitchModule } from 'ng-zorro-antd/switch';
import { NzListModule } from 'ng-zorro-antd/list';
import { NzRadioModule } from 'ng-zorro-antd/radio';
import { NzAlertModule } from 'ng-zorro-antd/alert';
import { NzLayoutModule } from 'ng-zorro-antd/layout';
import { NzMenuModule } from 'ng-zorro-antd/menu';
import { NzCollapseModule } from 'ng-zorro-antd/collapse';
import { NzInputModule } from 'ng-zorro-antd/input';
import { NzInputNumberModule } from 'ng-zorro-antd/input-number';
import { NzAutocompleteModule } from 'ng-zorro-antd/auto-complete';
import { NzCheckboxModule } from 'ng-zorro-antd/checkbox';
import { NzToolTipModule } from 'ng-zorro-antd/tooltip';
import { NzTabsModule } from 'ng-zorro-antd/tabs';
import { NzTableModule } from 'ng-zorro-antd/table';
import { NzTreeSelectModule } from 'ng-zorro-antd/tree-select';
import { NzGridModule } from 'ng-zorro-antd/grid';
import { NzPopoverModule } from 'ng-zorro-antd/popover';

import { IconsModule } from '../icons/icons.module';
import { NzAvatarModule } from 'ng-zorro-antd/avatar';

const AntdComponentModules = [
  NzButtonModule,
  NzInputModule,
  NzInputNumberModule,
  NzAutocompleteModule,
  NzCheckboxModule,
  NzFormModule,
  NzGridModule,
  NzTabsModule,
  NzSelectModule,
  NzModalModule,
  NzDropDownModule,
  NzSwitchModule,
  NzListModule,
  NzRadioModule,
  NzAlertModule,
  NzLayoutModule,
  NzMenuModule,
  NzCollapseModule,
  NzToolTipModule,
  NzTableModule,
  NzTreeSelectModule,
  NzPopoverModule,
  NzAvatarModule,
  // NgZorroAntdModule,
];

@NgModule({
  imports: [
    CommonModule,
    DragDropModule,
    TranslateModule.forRoot(),
    NgPipesModule,
    NguiAutoCompleteModule, // use antd
    MarkdownModule.forRoot(),
    ResizableModule,
    IconsModule,
    // ToastrModule.forRoot({
    //   newestOnTop: false,
    //   closeButton: true,
    //   positionClass: 'toast-top-center',
    //   enableHtml: true,
    //   countDuplicates: true,
    //   preventDuplicates: true,
    //   resetTimeoutOnDuplicate: true,
    // }),
    // HotToastModule.forRoot({
    //   position: 'top-center',
    //   reverseOrder: true,
    //   dismissible: true,
    //   closeStyle: {
    //     backgroundImage: `none`,
    //   },
    // }),
    ...AntdComponentModules,
  ],

  // DON'T FORGET TO EXPORT MODULE
  exports: [
    DragDropModule,
    TranslateModule,
    NgPipesModule,
    NguiAutoCompleteModule,
    MarkdownModule,
    ResizableModule,
    IconsModule,
    HotToastModule,
    ...AntdComponentModules,
  ],
})
export class SharedModule {
  static forRoot(): ModuleWithProviders<SharedModule> {
    return {
      ngModule: SharedModule,
      providers: [],
    };
  }
}
