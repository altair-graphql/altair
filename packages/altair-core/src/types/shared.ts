export type IDictionary<V = any> = Record<string, V>;

export type TODO = any;

// export type Omit<T, K extends keyof T> = Pick<T, Exclude<keyof T, K>>;

export interface BaseDocument {
  id: string;
  created_at?: number;
  updated_at?: number;
}

export interface BaseOwnableDocument extends BaseDocument {
  ownerUid: string;
}

type TypedOmit<T, K extends keyof T> = Omit<T, K>;
export type CreateDTO<
  T extends BaseDocument | BaseOwnableDocument
> = T extends BaseOwnableDocument
  ? TypedOmit<T, keyof BaseOwnableDocument>
  : TypedOmit<T, keyof BaseDocument>;
export type UpdateDTO<T extends BaseDocument | BaseOwnableDocument> = CreateDTO<
  T
> &
  Pick<T, 'id'>;
