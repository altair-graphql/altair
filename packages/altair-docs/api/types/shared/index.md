# types/shared

## Index

| Member | Description |
| :------ | :------ |
| [BaseDocument](interfaces/BaseDocument.md) | - |
| [BaseOwnableDocument](interfaces/BaseOwnableDocument.md) | - |
| [CreateDTO](type-aliases/CreateDTO.md) | - |
| [IDictionary](type-aliases/IDictionary.md) | - |
| [NoExtraProperties](type-aliases/NoExtraProperties.md) | - |
| [TODO](type-aliases/TODO.md) | - |
| [UpdateDTO](type-aliases/UpdateDTO.md) | - |
