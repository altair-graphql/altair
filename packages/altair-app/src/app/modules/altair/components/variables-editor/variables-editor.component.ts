import { Component, Input, Output, ViewChild, EventEmitter, OnChanges, ElementRef, SimpleChanges, AfterViewInit } from '@angular/core';

// Import the codemirror packages
import * as Codemirror from 'codemirror';
import 'codemirror/addon/hint/show-hint';
import 'codemirror/addon/lint/lint';
import 'codemirror/addon/fold/foldcode';
import 'codemirror/addon/fold/foldgutter';
import 'codemirror/addon/fold/brace-fold';
import 'codemirror/addon/fold/indent-fold';
// import 'codemirror/addon/display/autorefresh';
import 'codemirror/keymap/sublime';
import 'codemirror/mode/javascript/javascript';
import 'codemirror-graphql/variables/hint';
import 'codemirror-graphql/variables/lint';
import 'codemirror-graphql/variables/mode';
import { handleEditorRefresh } from '../../utils/codemirror/refresh-editor';
import { IDictionary } from '../../interfaces/shared';

const AUTOCOMPLETE_CHARS = /^[a-zA-Z0-9_\"\']$/;

@Component({
  selector: 'app-variables-editor',
  templateUrl: './variables-editor.component.html',
  styleUrls: ['./variables-editor.component.scss']
})
export class VariablesEditorComponent implements AfterViewInit, OnChanges {

  @Input() variables = '';
  @Input() variableToType: IDictionary = {};
  @Input() tabSize = 4;
  @Input() showVariableDialog = false;

  @Output() variablesChange = new EventEmitter();

  @ViewChild('editor', { static: true }) editor: ElementRef & { codeMirror: CodeMirror.Editor };

  variableEditorConfig = <any>{
    mode: 'graphql-variables',
    lineWrapping: true,
    lineNumbers: true,
    foldGutter: true,
    autoRefresh: true,
    dragDrop: false,
    tabSize: this.tabSize,
    indentUnit: this.tabSize,
    matchBrackets: true,
    autoCloseBrackets: true,
    keyMap: 'sublime',
    extraKeys: {
      'Ctrl-Enter': (cm: CodeMirror.Editor) => {},
      'Cmd-Enter': (cm: CodeMirror.Editor) => {},
      'Cmd-Space': (cm: any) => cm.showHint({ completeSingle: false }),
      'Ctrl-Space': (cm: any) => cm.showHint({ completeSingle: false }),
      'Alt-Space': (cm: any) => cm.showHint({ completeSingle: false }),
    },
    theme: 'default variable-editor mousetrap',
    gutters: ['CodeMirror-linenumbers', 'CodeMirror-foldgutter'],
    lint: {},
    hintOptions: {
      completeSingle: false
    },
  };

  constructor() { }

  ngAfterViewInit() {
    if (this.editor?.codeMirror) {
      (this.editor.codeMirror as any).on('keyup', (cm: CodeMirror.Editor, event: KeyboardEvent) => this.onKeyUp(cm, event));
    }
  }

  ngOnChanges(changes: SimpleChanges) {
    // Refresh the query result editor view when there are any changes
    // to fix any broken UI issues in it
    handleEditorRefresh(this.editor?.codeMirror);
    if (changes?.variableToType?.currentValue) {
      this.updateVariablesToType(changes.variableToType.currentValue);
    }

    if (changes?.tabSize?.currentValue) {
      this.variableEditorConfig.tabSize = this.tabSize;
      this.variableEditorConfig.indentUnit = this.tabSize;
    }

    if (changes?.activeWindowId?.currentValue || changes?.showVariableDialog?.currentValue) {
      handleEditorRefresh(this.editor?.codeMirror);
    }
  }

  

  onKeyUp(cm: CodeMirror.Editor, event: KeyboardEvent) {
    if (AUTOCOMPLETE_CHARS.test(event.key)) {
      this.editor.codeMirror.execCommand('autocomplete');
    }
  }

  updateVariablesToType(variableToType: IDictionary) {
    if (variableToType) {
      this.variableEditorConfig.lint.variableToType = variableToType;
      this.variableEditorConfig.hintOptions.variableToType = variableToType;
    }
  }

}
