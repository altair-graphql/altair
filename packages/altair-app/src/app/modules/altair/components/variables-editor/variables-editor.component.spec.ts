import { expect, describe, it } from '@jest/globals';

import { VariablesEditorComponent } from './variables-editor.component';
import { SharedModule } from '../../modules/shared/shared.module';
import { NgxTestWrapper, mount } from '../../../../../testing';
import { MockModule } from 'ng-mocks';
import { NO_ERRORS_SCHEMA } from '@angular/core';

describe('VariablesEditorComponent', () => {
  let wrapper: NgxTestWrapper<VariablesEditorComponent>;

  beforeEach(async () => {
    wrapper = await mount({
      component: VariablesEditorComponent,
      imports: [MockModule(SharedModule)],
      schemas: [NO_ERRORS_SCHEMA],
      propsData: {
        queryOperations: [],
      },
    });
  });

  it('should create', () => {
    expect(wrapper.componentInstance).toBeTruthy();
  });

  it('should render correctly', () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  it('should pass editor config to codemirror instance as options', () => {
    expect(wrapper.find('ngx-codemirror').props('options').mode).toBe(
      'graphql-variables'
    );
  });

  it('should update tabSize passed to codemirror when it changes', async () => {
    wrapper.setProps({
      tabSize: 2,
    });

    await wrapper.nextTick();
    expect(wrapper.find('ngx-codemirror').props('options').tabSize).toBe(2);

    wrapper.setProps({
      tabSize: 4,
    });

    await wrapper.nextTick();
    expect(wrapper.find('ngx-codemirror').props('options').tabSize).toBe(4);
  });
});
