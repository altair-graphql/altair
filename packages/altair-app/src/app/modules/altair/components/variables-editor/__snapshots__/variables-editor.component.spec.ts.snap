// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VariablesEditorComponent should render correctly 1`] = `
<app-variables-editor>
  
  <ngx-codemirror
    class="set-variable-textarea mousetrap"
  />
  
</app-variables-editor>
`;
